File: 7000/Baekjoon_No7569.py
from collections import deque

# 좌, 우, 하, 상, 아래, 위
dx = [1, -1, 0, 0, 0, 0]
dy = [0, 0, 1, -1, 0, 0]
dz = [0, 0, 0, 0, 1, -1]
queue = deque()
m, n, h = map(int, input().split())
matrix = [[list(map(int, input().split())) for _ in range(n)] for _ in range(h)]


def bfs():
    while queue:
        x, y, z = queue.popleft()
        for i in range(6):
            nx = x + dx[i]
            ny = y + dy[i]
            nz = z + dz[i]
            if 0 <= nx < h and 0 <= ny < n and 0 <= nz < m:
                if matrix[nx][ny][nz] == 0:
                    queue.append([nx, ny, nz])
                    matrix[nx][ny][nz] = matrix[x][y][z] + 1


# 이미 모든 토마토가 익어있는 경우
tmp = False
for i in range(h):
    for j in range(n):
        for k in range(m):
            if matrix[i][j][k] == 0:
                tmp = True
    if not tmp:
        print("0")
        exit()


# 토마토를 익혀봅시다
for i in range(h):
    for j in range(n):
        for k in range(m):
            if matrix[i][j][k] == 1:
                queue.append([i, j, k])
bfs()

# 흉년
for i in range(h):
    for j in range(n):
        for k in range(m):
            if matrix[i][j][k] == 0:
                print("-1")
                exit()

# 농사 잘 됐을때
result = 0
result = max(matrix[0][0])
for i in range(h):
    for j in range(n):
        if max(matrix[i][j]) > result:
            result = max(matrix[i][j])
print(result - 1)

"""
어제 풀었던 1차원 토마토 문제에서 진화한 3차원 토마토 문제이다
문제의 기본 틀은 변함 없지만 차원이 3차원으로 증가하여 리스트의 인덱스가 늘어나고
그에 맞게 여러 조건들을 추가해주어야 했다. 어렵진 않았는데 중간에 오타를 못봐서
시간이 오래 걸렸던 것 같다.
"""

File: 7000/Baekjoon_No7569.java
import java.io.*;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

class Tomato {
    int row;
    int col;
    int height;

    public Tomato(int height, int row, int col) {
        this.row = row;
        this.col = col;
        this.height = height;
    }
}

public class b7576 {
    static int rowArr[] = {-1, 0, 1, 0, 0, 0};
    static int colArr[] = {0, 1, 0, -1, 0, 0};
    static int heightArr[] = {0, 0, 0, 0, 1, -1};
    static int M, N, H;
    static int tomato[][][];
    static int result;
    static Queue<Tomato> queue = new LinkedList<>();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());
        M = Integer.parseInt(st.nextToken());
        N = Integer.parseInt(st.nextToken());
        H = Integer.parseInt(st.nextToken());
        tomato = new int[H + 1][N + 1][M + 1];

        for (int i = 1; i <= H; i++) {
            for (int j = 1; j <= N; j++) {
                st = new StringTokenizer(br.readLine());
                for (int k = 1; k <= M; k++) {
                    tomato[i][j][k] = Integer.parseInt(st.nextToken());
                    if (tomato[i][j][k] == 1)
                        queue.add(new Tomato(i, j, k));
                }
            }
        }

        bw.write(bfs() + "\n");
        bw.flush();
        bw.close();
        br.close();
    }

    public static int bfs() {
        while (!queue.isEmpty()) {
            Tomato point = queue.poll();

            int row = point.row;
            int col = point.col;
            int height = point.height;

            for (int i = 0; i < 6; i++) {
                int newHeight = height + heightArr[i];
                int newRow = row + rowArr[i];
                int newCol = col + colArr[i];
                if (checkPoint(newHeight, newRow, newCol)) {
                    queue.add(new Tomato(newHeight, newRow, newCol));
                    tomato[newHeight][newRow][newCol] = tomato[height][row][col] + 1;
                }
            }
        }

        int result = Integer.MIN_VALUE;

        for (int i = 1; i <= H; i++) {
            for (int j = 1; j <= N; j++) {
                for (int k = 1; k <= M; k++) {
                    if (tomato[i][j][k] == 0) {
                        return -1;
                    }
                    result = Math.max(result, tomato[i][j][k]);
                }
            }
        }
        if (result == 1) {
            return 0;
        } else {
            return result - 1;
        }
    }

    public static boolean checkPoint(int height, int row, int col) {
        if (height < 1 || height > H || row < 1 || row > N || col < 1 || col > M)
            return false;
        if (tomato[height][row][col] == 0)
            return true;
        else
            return false;
    }
}
